src/hazelcore/semantics/dynamics/Dynamics.re

```diff
@@ -4,72 +4,113 @@ open Sexplib.Std;
+    module Let_syntax = {
+      module Let_syntax = {
+        let bind =
+            (x: t, ~f: ((DHPat.t, HTyp.t, Contexts.t, Delta.t)) => t): t =>
+          switch (x) {
+          | DoesNotExpand => DoesNotExpand
+          | Expands(dp, ty, ctx, delta) => f((dp, ty, ctx, delta))
+          };
+      };
+    };
+  };
 
   let rec syn_expand =
-          (ctx: Contexts.t, delta: Delta.t, p: UHPat.t): expand_result =>
+          (ctx: Contexts.t, delta: Delta.t, p: UHPat.t): ExpandResult.t =>
     syn_expand_opseq(ctx, delta, p)
   and syn_expand_opseq =
       (ctx: Contexts.t, delta: Delta.t, OpSeq(skel, seq): UHPat.opseq)
-      : expand_result =>
+      : ExpandResult.t =>
     syn_expand_skel(ctx, delta, skel, seq)
   and syn_expand_skel =
       (ctx: Contexts.t, delta: Delta.t, skel: UHPat.skel, seq: UHPat.seq)
-      : expand_result =>
+      : ExpandResult.t =>
     switch (skel) {
     | Placeholder(n) =>
       syn_expand_operand(ctx, delta, seq |> Seq.nth_operand(n))
     | BinOp(InHole(TypeInconsistent as reason, u), op, skel1, skel2)
     | BinOp(InHole(WrongLength as reason, u), Comma as op, skel1, skel2) =>
       let skel_not_in_hole = Skel.BinOp(NotInHole, op, skel1, skel2);
       switch (syn_expand_skel(ctx, delta, skel_not_in_hole, seq)) {
-      | DoesNotExpand => DoesNotExpand
+      | ExpandResult.DoesNotExpand => ExpandResult.DoesNotExpand
       | Expands(dp, _, ctx, delta) =>
         let gamma = Contexts.gamma(ctx);
         let delta =
           MetaVarMap.extend_unique(delta, (u, (PatternHole, Hole, gamma)));
         Expands(NonEmptyHole(reason, u, 0, dp), Hole, ctx, delta);
       };
-    | BinOp(InHole(WrongLength, _), _, _, _) => DoesNotExpand
-    | BinOp(NotInHole, Comma, skel1, skel2) =>
-      switch (syn_expand_skel(ctx, delta, skel1, seq)) {
-      | DoesNotExpand => DoesNotExpand
-      | Expands(dp1, ty1, ctx, delta) =>
-        switch (syn_expand_skel(ctx, delta, skel2, seq)) {
-        | DoesNotExpand => DoesNotExpand
-        | Expands(dp2, ty2, ctx, delta) =>
-          let dp = DHPat.Pair(dp1, dp2);
-          Expands(dp, Prod(ty1, ty2), ctx, delta);
-        }
+    | BinOp(InHole(WrongLength, _), _, _, _) => ExpandResult.DoesNotExpand
+    | BinOp(NotInHole, Comma, _, _) =>
+      switch (UHPat.get_tuple_elements(skel)) {
+      | [skel1, skel2, ...tail] =>
+        let%bind.ExpandResult (dp1, ty1, ctx, delta) =
+          syn_expand_skel(ctx, delta, skel1, seq);
+        let%bind.ExpandResult (dp2, ty2, ctx, delta) =
+          syn_expand_skel(ctx, delta, skel2, seq);
+        tail
+        |> ListUtil.map_with_accumulator_opt(
+             ((dp_acc, ctx, delta), skel) => {
+               syn_expand_skel(ctx, delta, skel, seq)
+               |> ExpandResult.to_option
+               |> Option.map(((dp, ty, ctx, delta)) =>
+                    ((DHPat.Pair(dp_acc, dp), ctx, delta), ty)
+                  )
+             },
```

Comment:
hard to test the rest of this until the reverse bug in the comment is fixed, so ping me when that's done